checker/src/main/java/org/checkerframework/checker/pico/PICONoInitVisitor.java

```diff
@@ -0,0 +1,588 @@
+    /** whether to only check observational purity */
+    protected final boolean abstractStateOnly;
+
+    /**
+     * Create a new PICONoInitVisitor.
+     *
+     * @param checker the checker
+     */
+    public PICONoInitVisitor(BaseTypeChecker checker) {
+        super(checker);
+        abstractStateOnly = checker.hasOption("abstractStateOnly");
+    }
+
+    @Override
+    protected TypeValidator createTypeValidator() {
+        return new PICOValidator(checker, this, atypeFactory);
+    }
+
+    @Override
+    protected void checkConstructorResult(
+            AnnotatedExecutableType constructorType, ExecutableElement constructorElement) {}
+
+    // This method is for validating usage of mutability qualifier is conformable to element
+    // declaration,
+    // Ugly thing here is that declarationType is not the result of calling the other method -
+    // PICOTypeUtil#getBoundTypeOfTypeDeclaration. Instead it's the result of calling
+    // ATF#getAnnotatedType(Element).
+    // Why it works is that PICOTypeUtil#getBoundTypeOfTypeDeclaration and
+    // ATF#getAnnotatedType(Element) has
+    // the same effect most of the time except on java.lang.Object. We need to be careful when
+    // modifying
+    // PICOTypeUtil#getBoundTypeOfTypeDeclaration so that it has the same behaviour as
+    // ATF#getAnnotatedType(Element)
+    // (at least for types other than java.lang.Object)
+    @Override
+    public boolean isValidUse(
+            AnnotatedDeclaredType declarationType, AnnotatedDeclaredType useType, Tree tree) {
+
+        // FIXME workaround for poly anno, remove after fix substitutable poly and add poly vp rules
+        if (useType.hasAnnotation(atypeFactory.POLY_MUTABLE)) {
+            return true;
+        }
+
+        AnnotationMirror declared = declarationType.getAnnotationInHierarchy(atypeFactory.READONLY);
+        AnnotationMirror used = useType.getAnnotationInHierarchy(atypeFactory.READONLY);
+
+        return isAdaptedSubtype(used, declared);
+    }
+
+    @Override
+    public boolean isValidUse(AnnotatedArrayType type, Tree tree) {
+        // You don't need adapted subtype if the decl bound is guaranteed to be RDM.
+        // That simply means that any use is valid except bottom.
+        AnnotationMirror used = type.getAnnotationInHierarchy(atypeFactory.READONLY);
+        return !AnnotationUtils.areSame(used, atypeFactory.BOTTOM);
+    }
+
+    /**
+     * Check if the lhs is adapted subtype of rhs.
+     *
+     * @param lhs the lhs annotation
+     * @param rhs the rhs annotation
+     * @return true if lhs is adapted subtype of rhs, false otherwise
+     */
+    private boolean isAdaptedSubtype(AnnotationMirror lhs, AnnotationMirror rhs) {
+        PICOViewpointAdapter vpa = atypeFactory.getViewpointAdapter();
+        AnnotationMirror adapted = vpa.combineAnnotationWithAnnotation(lhs, rhs);
+        return atypeFactory.getQualifierHierarchy().isSubtypeQualifiersOnly(adapted, lhs);
+    }
```

Comment:
After #1032, this method and the override of `isValidUse(
            AnnotatedDeclaredType declarationType, AnnotatedDeclaredType useType, Tree tree)` can be removed.